From e87687d72688e980b929920b7d77dca26fff169e Mon Sep 17 00:00:00 2001
From: Alistair Francis <alistair.francis@wdc.com>
Date: Mon, 21 Aug 2023 14:00:46 -0400
Subject: [PATCH] cryptlib_openssl: ecd: Allow disabling code

The OpenSSL source code describes the crypto include as:
"Internal EC functions for other submodules: not for application use"
 - https://github.com/openssl/openssl/blob/master/include/crypto/ec.h

Using the internal APIS makes it difficult to use libspdm as a library
with other packages. So let's remove the uses of the internal API and
instead use the public API.

The current ECD code uses internal APIs, making it unsuitable for use in
production code or libraries.

The supported way to do this is via OSSL params, either with
EVP_PKEY_fromdata() [1] or using EVP_PKEY_set_octet_string_param().

Unfortunately this isn't supported in OpenSSL and ed25519_set_params()
and ed448_set_params() will always return 1, indicating no support.

As there doesn't appear to be a supported method in OpenSSL to set the
public and private keys, let's instead allow users to disable this
support so the library can be used with the regular OpenSSL libraries.

https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_fromdata.html

Signed-off-by: Alistair Francis <alistair.francis@wdc.com>
Upstream: https://github.com/DMTF/libspdm/commit/e87687d72688e980b929920b7d77dca26fff169e
---
 os_stub/cryptlib_openssl/pk/ecd.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/os_stub/cryptlib_openssl/pk/ecd.c b/os_stub/cryptlib_openssl/pk/ecd.c
index d7cc156d86..23dbd0390b 100644
--- a/os_stub/cryptlib_openssl/pk/ecd.c
+++ b/os_stub/cryptlib_openssl/pk/ecd.c
@@ -12,6 +12,9 @@
  **/
 
 #include "internal_crypt_lib.h"
+
+#if (LIBSPDM_EDDSA_ED25519_SUPPORT) || (LIBSPDM_EDDSA_ED448_SUPPORT)
+
 #include <openssl/evp.h>
 #include <crypto/evp.h>
 
@@ -471,3 +474,4 @@ bool libspdm_eddsa_verify(const void *ecd_context, size_t hash_nid,
     EVP_MD_CTX_free(ctx);
     return true;
 }
+#endif /* (LIBSPDM_EDDSA_ED25519_SUPPORT) || (LIBSPDM_EDDSA_ED448_SUPPORT) */
-- 
2.40.1

